[Skia] Implement font collection / fragment identifiers

dpino opened the pull request:
#### ab9b9abab22f7a26d56d5c81ee9911cdb2506064
<pre>
[Skia] Implement font collection / fragment identifiers
<a href="https://bugs.webkit.org/show_bug.cgi?id=186086">https://bugs.webkit.org/show_bug.cgi?id=186086</a>

Reviewed by Carlos Garcia Campos.

When loading a specific font from a  font collection, iterate over all
the available fonts in the font collection until finding the desired
font.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/svg/custom/glyph-transformation-with-hkern-expected.txt:
* LayoutTests/platform/glib/svg/custom/svg-fonts-without-missing-glyph-expected.txt:
* LayoutTests/platform/glib/svg/foreignObject/text-tref-02-b-expected.txt:
* LayoutTests/platform/glib/svg/text/text-overflow-ellipsis-svgfont-expected.txt:
* LayoutTests/platform/glib/svg/text/text-overflow-ellipsis-svgfont-kerning-ligatures-expected.txt: Removed.
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1-SE/paths-dom-02-f-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1-SE/types-dom-04-b-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/filters-light-04-f-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/filters-turb-02-f-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/fonts-elem-03-b-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/fonts-elem-04-b-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/fonts-elem-07-b-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-intro-01-t-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-intro-02-b-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-intro-03-b-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-intro-04-t-expected.txt:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1-SE/paths-dom-02-f-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1-SE/types-dom-04-b-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/filters-light-04-f-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/filters-turb-02-f-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/fonts-elem-03-b-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/fonts-elem-04-b-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/fonts-elem-07-b-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/text-intro-01-t-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/text-intro-02-b-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/text-intro-03-b-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/text-intro-04-t-expected.txt:
* Source/WebCore/platform/graphics/skia/FontCustomPlatformDataSkia.cpp:
(WebCore::FontCustomPlatformData::create):

Canonical link: <a href="https://commits.webkit.org/299380@main">https://commits.webkit.org/299380@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2745acbd56d2bca5dcff5f5167e93c1cbc204f10

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118844 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38525 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29176 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/125033 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70906 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120722 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39221 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47107 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/125033 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121797 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/31260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106551 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/125033 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/30322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24665 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68692 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/100703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24852 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128084 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45751 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/34550 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/98851 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46117 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102771 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98631 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44077 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22082 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42310 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18929 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45621 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51299 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45086 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48431 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46771 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

iangrunert commented on LayoutTests/platform/glib/TestExpectations:
```diff
@@ -4841,6 +4836,10 @@ imported/w3c/web-platform-tests/selection/caret/move-around-generated-content.ht
 
+# Now that partial fonts are working, two characters previously rendered as a square are now showing correctly, which makes the test fail. 
+# Probably the test should be updated.
+fast/text/text-combine-placement.html [ Failure ]
```
Can you add this to the windows TestExpectations as well? Fails there now for what looks like similar reasons.

dpino replied:
Ok. Since Windows seems to get the same results as WebKitGTK/WPEWebkit, I also took advantage to remove `fast/text/font-collection.html` and `fast/text/font-collection-2.html` from Windows' TestExpectations.

iangrunert replied:
Thanks, I appreciate it!

dpino committed:
[Skia] Implement font collection / fragment identifiers
https://bugs.webkit.org/show_bug.cgi?id=186086

Reviewed by Carlos Garcia Campos.

When loading a specific font from a  font collection, iterate over all
the available fonts in the font collection until finding the desired
font.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/svg/custom/glyph-transformation-with-hkern-expected.txt:
* LayoutTests/platform/glib/svg/custom/svg-fonts-without-missing-glyph-expected.txt:
* LayoutTests/platform/glib/svg/foreignObject/text-tref-02-b-expected.txt:
* LayoutTests/platform/glib/svg/text/text-overflow-ellipsis-svgfont-expected.txt:
* LayoutTests/platform/glib/svg/text/text-overflow-ellipsis-svgfont-kerning-ligatures-expected.txt: Removed.
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1-SE/paths-dom-02-f-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1-SE/types-dom-04-b-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/filters-light-04-f-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/filters-turb-02-f-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/fonts-elem-03-b-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/fonts-elem-04-b-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/fonts-elem-07-b-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-intro-01-t-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-intro-02-b-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-intro-03-b-expected.txt:
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-intro-04-t-expected.txt:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1-SE/paths-dom-02-f-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1-SE/types-dom-04-b-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/filters-light-04-f-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/filters-turb-02-f-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/fonts-elem-03-b-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/fonts-elem-04-b-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/fonts-elem-07-b-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/text-intro-01-t-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/text-intro-02-b-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/text-intro-03-b-expected.txt:
* LayoutTests/platform/wpe/svg/W3C-SVG-1.1/text-intro-04-t-expected.txt:
* Source/WebCore/platform/graphics/skia/FontCustomPlatformDataSkia.cpp:
(WebCore::FontCustomPlatformData::create):

Canonical link: https://commits.webkit.org/299380@main
